File: flow-client/src/main/java/com/vaadin/client/UILifecycle.java
/*
 * Copyright 2000-2026 Vaadin Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */
package com.vaadin.client;

import jsinterop.annotations.JsFunction;
import jsinterop.annotations.JsMethod;
import jsinterop.annotations.JsOverlay;
import jsinterop.annotations.JsType;

/**
 * Manages the lifecycle of a UI. Pure {@code @JsType(isNative=true)} binding to
 * the TypeScript implementation at
 * {@code src/main/frontend/internal/client/UILifecycle.ts}. The Java
 * {@link UIState} enum is retained for type-safe consumer code; the JS side
 * stores the matching string and {@link #getState()} / {@link #setState} adapt
 * between the two on each call.
 *
 * @author Vaadin Ltd
 * @since 1.0
 */
@JsType(isNative = true, namespace = "Vaadin.Flow.internal.client", name = "UILifecycle")
public class UILifecycle {

    /** State of the UI. */
    public enum UIState {
        INITIALIZING, RUNNING, TERMINATED;
    }

    /** JS handler-registration shape returned by {@link #addHandler}. */
    @JsType(isNative = true)
    public interface HandlerRegistration {
        void removeHandler();
    }

    /** Native event shape passed to {@link StateChangeHandler}. */
    @JsType(isNative = true)
    public interface StateChangeEvent {
        UILifecycle getUiLifecycle();
    }

    /** Listener for UI state changes. */
    @FunctionalInterface
    @JsFunction
    @SuppressWarnings("unusable-by-js")
    public interface StateChangeHandler {
        void onUIStateChanged(StateChangeEvent event);
    }

    public UILifecycle() {
        // Defined by the TS class constructor.
    }

    @JsMethod(name = "getStateName")
    native String getStateNameImpl();

    @JsMethod(name = "setStateName")
    native void setStateNameImpl(String state);

    public native boolean isRunning();

    public native boolean isTerminated();

    public native HandlerRegistration addHandler(StateChangeHandler handler);

    /** Gets the current UI state. */
    @JsOverlay
    public final UIState getState() {
        return UIState.valueOf(getStateNameImpl());
    }

    /** Sets the UI state. Only forward transitions are allowed. */
    @JsOverlay
    public final void setState(UIState state) {
        setStateNameImpl(state.name());
    }
}

File: flow-client/src/main/frontend/internal/UILifecycle.ts
/*
 * Copyright 2000-2026 Vaadin Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */

// TypeScript port of com.vaadin.client.UILifecycle, built alongside the Java
// version. It manages a UI's lifecycle state, which may only advance forward
// INITIALIZING -> RUNNING -> TERMINATED, firing a state-change event on each
// transition. The GWT EventBus is replaced by a listener set.

import type { EventRemover } from './reactive/reactive';

/** The lifecycle state of a UI; the order defines the allowed forward transitions. */
export const UIState = {
  INITIALIZING: 'INITIALIZING',
  RUNNING: 'RUNNING',
  TERMINATED: 'TERMINATED'
} as const;

export type UIState = (typeof UIState)[keyof typeof UIState];

// Ordinals matching the Java enum order, used to enforce single-step forward
// transitions.
const ORDINAL: Record<UIState, number> = {
  INITIALIZING: 0,
  RUNNING: 1,
  TERMINATED: 2
};

/** Event fired when the lifecycle state changes; mirrors UILifecycle.StateChangeEvent. */
export interface StateChangeEvent {
  getUiLifecycle(): UILifecycle;
}

/** A listener for UI lifecycle state changes; mirrors StateChangeHandler. */
export type StateChangeHandler = (event: StateChangeEvent) => void;

/** Manages the lifecycle state of a UI; mirrors UILifecycle.java. */
export class UILifecycle {
  private state: UIState = UIState.INITIALIZING;

  private readonly handlers = new Set<StateChangeHandler>();

  /** The current lifecycle state. */
  getState(): UIState {
    return this.state;
  }

  /**
   * Advances the state. Only single-step forward transitions
   * (INITIALIZING -> RUNNING -> TERMINATED) are allowed; fires a state-change
   * event.
   */
  setState(state: UIState): void {
    if (ORDINAL[state] !== ORDINAL[this.state] + 1) {
      throw new Error(`Tried to move from state ${this.state} to ${state} which is not allowed`);
    }
    this.state = state;
    const event: StateChangeEvent = { getUiLifecycle: () => this };
    this.handlers.forEach((handler) => handler(event));
  }

  /** Whether the state is RUNNING. */
  isRunning(): boolean {
    return this.state === UIState.RUNNING;
  }

  /** Whether the state is TERMINATED. */
  isTerminated(): boolean {
    return this.state === UIState.TERMINATED;
  }

  /** Adds a state-change handler, returning a remover. */
  addHandler(handler: StateChangeHandler): EventRemover {
    this.handlers.add(handler);
    return {
      remove: () => {
        this.handlers.delete(handler);
      }
    };
  }
}
